### ISSUE TITLE
[discussion] handling debug output

### ISSUE THREAD

**Issue by jgibbon** [Wunderfitz/harbour-fernschreiber#174]
This isn't really high priority for now, but I'd like to hear your thoughts on handling debug output, because at some point before 1.0 we should adress this IMHO.

- Should it only be available in debug builds?
- Should it be accessible via command line parameter?
- (How) do we handle QML output (console)?

The first option would result in a "cleaner" release build, whereas the second could aid users helping out with bug hunting. 

Basically the same goes for QML – sometimes it's useful, but it's mostly good for debugging, which most users won't ever do. Also I'm not aware of a (sane) way to get rid of QML/JS parts at compile time. Handling it at runtime would mean a bit of performance penalty even for "not using" it.

What do you think?

cheers!

**Comment by monich** [comment]
Regarding `qDebug()`, I prefer to completely strip debug output in release build with preprocessor, to avoid any runtime overhead and reduce the size of the executable.

If debug output is left in the code, then it's better to use [logging categories](https://doc.qt.io/qt-5/qloggingcategory.html) for that. Those can be turned on at run time one by one. It's not as easy as passing a command line option ([especially](https://bugreports.qt.io/browse/QTBUG-43293) if you need to enable more than one) but it's possible via environment variable or a config file. Documentation even claims that disabled logging categories don't have any runtime overhead, although I don't quite understand how that could be achieved (specifically, how to avoid evaluation of run time arguments).

Or it could be `qDebug()` in debug build and categories in release.

With these LOG macros that I've been introducing, tweaking debug logging should be a fairly trivial exercise. As soon as it's clear how it should be handled.

**Comment by monich** [comment]
And I'm not aware of any adequate solution for QML.

One could check some debug log flag before every call to `console.log()`. That would look pretty ugly but it's probably the best solution if you want to leave that debug output there. If that flag is off, the runtime overhead is reduced to parsing/compiling extra QML code and checking the flag every time when you get to those statements. The former is probably the worst part.

Alternatively, one could make a function which takes a string and checks a flag or something before actually printing it, but parameters would still be evaluated, formatted, string constructed and passed to that function, the flag can only be checked at the very end of that process. All the runtime overhead will be there no matter what.

**Comment by Wunderfitz** [comment]
I would agree - let's try stripping all debug log outputs from release builds - the work @monich has been doing so far is really awesome and we can continue doing that once we touch code that still contains qDebug() statements. 

For QML we could introduce an own loggin method in `Functions` or somewhere else which can replace all calls to `console.log` completely and that includes the check if a debug flag is set. By that we wouldn't spoil the QML/JS code and get the same result.

**Comment by monich** [comment]
> For QML we could introduce an own loggin method in `Functions` or somewhere else which can replace all calls to `console.log` completely and that includes the check if a debug flag is set. By that we wouldn't spoil the QML/JS code and get the same result.

but wouldn't eliminate(most of) runtime overhead. Not sure how big it is, though.

**Comment by monich** [comment]
Just noticed that the build which I prepare myself in OBS doesn't produce any `qDebug()` output, only QML console logs (and a few things from tdlib which I also built in OBS).

**Comment by jgibbon** [comment]
Since JS is a bit like chewing gum, maybe we can get away with checking the debug flag once and replacing the whole function with void if it is not present... 

Not sure how that would work out in the QML import context, though, but it could work.

**Comment by jgibbon** [comment]
Agreed, but that's kind of the point: 
For a small check at application start (plus handling the flag in js on initialization), we could basically skip all checks further down the line in favour of running an empty function – apart from stripping the code completely, of course – is basically as cheap as it can get.

Of course this idea only applies to the case that we decide to keep frontend logging for debug builds or via command line flag, but it could possibly be the fastest one for that case (provided there's a few "console.log" calls at all). And it would have to be tried out, of course.

To be clear, I'm not at all against dropping frontend logging by default, just trying to formulate a more performant alternative to "keep and check with every call".

**Comment by monich** [comment]
A command line option (or an environment variable) can be exposed to QML as a context property.

**Comment by jgibbon** [comment]
Yes, that was implicitly included in the "small check at application start". 
One could also default to the "`void()`" part in JS and only explicitly redefine it via `invokeMethod` from c++ when needed/wanted.

**Comment by monich** [comment]
a `console.log()` replacement could look like this:
```JavaScript
.pragma library

var enabled = false

function log(arg1,arg2,arg3,arg4,arg5,arg6,arg7,arg8,arg9) {
    if (enabled) {
        if (arg2 === undefined) {
            console.log(arg1)
        } else if (arg3 === undefined) {
            console.log(arg1,arg2)
        } else if (arg4 === undefined) {
            console.log(arg1,arg2,arg3)
        } else if (arg5 === undefined) {
            console.log(arg1,arg2,arg3,arg4)
        } else if (arg6 === undefined) {
            console.log(arg1,arg2,arg3,arg4,arg5)
        } else if (arg7 === undefined) {
            console.log(arg1,arg2,arg3,arg4,arg5,arg6)
        } else if (arg8 === undefined) {
            console.log(arg1,arg2,arg3,arg4,arg5,arg6,arg7)
        } else if (arg9 === undefined) {
            console.log(arg1,arg2,arg3,arg4,arg5,arg6,arg7,arg8)
        } else {
            console.log(arg1,arg2,arg3,arg4,arg5,arg6,arg7,arg8,arg9)
        }
    }
}
```
```QML
import "debug.js" as Debug

...
Debug.log("this","is","some","output")
```

**Comment by monich** [comment]
or like this ))
```JavaScript
.pragma library

var enabled = true

function log(arg1,arg2,arg3,arg4,arg5,arg6,arg7,arg8,arg9) {
    if (enabled) {
        if (arg9 === undefined) {
            if (arg8 === undefined) {
                if (arg7 === undefined) {
                    if (arg6 === undefined) {
                        if (arg5 === undefined) {
                            if (arg4 === undefined) {
                                if (arg3 === undefined) {
                                    if (arg2 === undefined) {
                                        console.log(arg1)
                                    } else {
                                        console.log(arg1,arg2)
                                    }
                                } else {
                                    console.log(arg1,arg2,arg3)
                                }
                            } else {
                                console.log(arg1,arg2,arg3,arg4)
                            }
                        } else {
                            console.log(arg1,arg2,arg3,arg4,arg5)
                        }
                    } else {
                        console.log(arg1,arg2,arg3,arg4,arg5,arg6)
                    }
                } else {
                    console.log(arg1,arg2,arg3,arg4,arg5,arg6,arg7)
                }
            } else {
                console.log(arg1,arg2,arg3,arg4,arg5,arg6,arg7,arg8)
            }
        } else {
            console.log(arg1,arg2,arg3,arg4,arg5,arg6,arg7,arg8,arg9)
        }
    }
}
```
that allows inserting `undefined` into the middle.

**Comment by jgibbon** [comment]
That looks… interesting. I was thinking more among the lines of
```
.pragma library
function enable() {
  log = function() {
    console.log.apply(console, arguments);
  }
};
var log = void;
```

**Comment by jgibbon** [comment]
Depending on how the engine handles the calling scope, one could possibly even get away with
```JavaScript
.pragma library
function enable() {
  log = console.log;
};
var log = void;
```

**Comment by monich** [comment]
The ones I pasted would prefix every line with something dull and boring like `[D] log:17 - ` which is useless. It would be much nicer to automatically insert the name of the component which is doing the logging but I'm not sure if that's possible at all. Could anything interesting be pulled out of QML engine context if that logging function is moved to the native code...?

**Comment by monich** [comment]
`console.log()` implementation could actually be an example of what can be done on the native side. Although it's probably some kind of special-cased builtin internal thing which has more powers than a regular function.

**Comment by jgibbon** [comment]
Now that I've actually tried it instead of blindly typing things into a comment field, the following would actually work and retain the calling method name like `[D] onTriggered:620 - debug output`:

```JavaScript
.pragma library

var log = console.log.bind(console);
function disable() {
    log = function(){};
}
```
The QML counterpart would still be something like
```QML
import "../js/debug.js" as Debug
Item {
    Component.onCompleted: {
        Debug.log("now you see me");
        Debug.disable();
        Debug.log("now you don't");
    }
}
```
which results in something like
```
[D] onCompleted:4 - now you see me
```

**Comment by monich** [comment]
Shamanism :) I don't think I even understand this `log = console.log.bind(console)` spell. Must be magic :)

**Comment by jgibbon** [comment]
That's how I regularly feel reading c++, tbh, but it's getting a tiny bit better at least. ;)

Obviously, the JS would make more sense defaulting to "empty function" and manually enabling it… but that change is probably easier than typing this sentence.

So I guess we have basic procedures in mind (well, I'd be doing a sub par job implementing the c++ part, but you know what I mean). Nice.

Should we just start a branch here where we split the work? 
I volunteer for the boring qml/js part including replacements + imports and could even handle an "invokeMethod" (shocking, I know!) from c++ ;)

**Comment by monich** [comment]
I suggest to do QML logging changes first and then native LOGs as a separate effort (probably via categories, it's silly to strip those if QML logs are staying).

**Comment by jgibbon** [comment]
OK, I should have some time to prepare the branch with that this weekend.

**Comment by monich** [comment]
Some food for thought.

I played with C++ <=> JavaScript interaction a bit and this is what I came up with:

JavaScript:
```JavaScript
.pragma library
.import WerkWolf.Fernschreiber 1.0 as Fernschreiber
var log = Fernschreiber.DebugLog.enabled ? console.log : function(){}
```
C++ class:
```C++
class DebugLog : public QObject
{
    Q_OBJECT
    Q_PROPERTY(bool enabled READ isEnabled WRITE setEnabled NOTIFY enabledChanged)
public:
    DebugLog(QObject* parent = Q_NULLPTR) : QObject(parent), enabled(false) {}
    static QObject* createSingleton(QQmlEngine* aEngine, QJSEngine* aScript) { return new DebugLog(); }
    bool isEnabled() const { return enabled; }
    void setEnabled(bool value) {
        if (enabled != value) {
            enabled = value;
            Q_EMIT enabledChanged(value);
        }
    }
Q_SIGNALS:
    void enabledChanged(bool value);
private:
    bool enabled;
};
```
registered with QML runtime like this:
```C++
qmlRegisterSingletonType<DebugLog>("WerkWolf.Fernschreiber", 1, 0, "DebugLog", DebugLog::createSingleton);
```

**Comment by monich** [comment]
and this magic spell:
```JavaScript
(function() { Fernschreiber.DebugLog.enabledChanged.connect(function(value) {
log = value ? console.log : function(){}
})})();
```
would even allow the JavaScript module to react to `Fernschreiber.DebugLog.enabled` changes at run time. Which is probably unnecessary.

**Comment by jgibbon** [comment]
That looks viable.
Should I just copy your Class as a base to work on when starting the branch?

**Comment by monich** [comment]
> That looks viable.
> Should I just copy your Class as a base to work on when starting the branch?

Do whatever, it's just an example, I wasn't sure if it's going to be useful at all!

**Comment by jgibbon** [comment]
I just want to make sure it doesn't interfere with the rest of c++ logging changes you had in mind in case this one should manage both parts in one place.

**Comment by jgibbon** [comment]
As announced, I've prepared that branch (including draft PR #185) with the qml/js part and would like to hand over the torch to @monich for now :)

**Comment by monich** [comment]
> Documentation even claims that disabled logging categories don't have any runtime overhead, although I don't quite understand how that could be achieved (specifically, how to avoid evaluation of run time arguments).

Hehe, quick look at `qloggingcategory.h` reveals how qCxxx macros avoid runtime penalty:

```C++
#define qCDebug(category, ...) \
    for (bool qt_category_enabled = category().isDebugEnabled(); qt_category_enabled; qt_category_enabled = false) \
        QMessageLogger(QT_MESSAGELOG_FILE, QT_MESSAGELOG_LINE, QT_MESSAGELOG_FUNC, category().categoryName()).debug(__VA_ARGS__)
```

**Comment by monich** [comment]
Submitted https://github.com/Wunderfitz/harbour-fernschreiber/pull/191 for the C++ part.
